# importing libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
data_path=r'/Users/Lilian/Downloads/Data Analytics Course/Data Analytics Immersion/Acheivement 4/04-2024 Instacart Basket Analysis/'

In [3]:
df_ords= pd.read_csv(os.path.join(data_path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [4]:
df_prods = pd.read_csv(os.path.join(data_path,'02 Data','Original Data','products.csv'),index_col = False)

In [5]:
#df_dep= pd.read_csv(os.path.join(data_path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [10]:
# importing additional datasets for final project

In [6]:
df_prods_prior = pd.read_csv(os.path.join(data_path,'02 Data','Original Data','orders_products_prior.csv'),index_col = False)

In [16]:
df_customers = pd.read_csv(os.path.join(data_path,'02 Data','Original Data','customers.csv'),index_col = False)

# Data Consistency Checks

### 1. Mixed-Type Data

In [47]:
# create a dataframe
df_test = pd.DataFrame()

In [48]:
# create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [49]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [50]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [51]:
df_test['mix'] = df_test['mix'].astype('str')

### 2. Missing Values

In [52]:
# finding missing values in products data frame
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [53]:
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [54]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
df_prods.shape

(49693, 5)

In [55]:
# creating a new dataframe which doesn't contain the null product name rows from df_prods.
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [56]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [57]:
#checking for full row duplicates  by creating a new subset of dataframe which contains only duplicate rows
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [58]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [59]:
df_prods_clean.shape

(49677, 5)

In [60]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [61]:
df_prods_clean_no_dups.shape

(49672, 5)

In [69]:
# Exporting cleaned up product dataframe 
df_prods_clean_no_dups.to_csv(os.path.join(data_path, '02 Data','Prepared Data', 'products_cleaned_up.csv'),index = False)

In [68]:
df_prods_clean_no_dups.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices'], dtype='object')

## Data checks for orders table for final project

In [64]:
df_prods_clean_no_dups.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices'], dtype='object')

In [23]:
# repeating steps for orders df for final project - check for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [24]:
# checking for duplicate values on orders df and saving them into a new df 
df_dup_ords = df_ords[df_ords.duplicated()]

In [25]:
df_dup_ords.shape

(0, 7)

### consistency checks for df_prods_prior

In [7]:
df_prods_prior.isnull().sum()

order_id             0
product_id           0
add_to_cart_order    0
reordered            0
dtype: int64

In [8]:
df_prods_prior_dups = df_prods_prior[df_prods_prior.duplicated()]

In [9]:
df_prods_prior_dups.shape

(0, 4)

### consistency checks for customers

In [17]:
df_customers.isnull().sum()

user_id             0
First Name      11259
Surnam              0
Gender              0
STATE               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [18]:
df_customers_dups = df_customers[df_customers.duplicated()]

In [19]:
df_customers_dups.shape

(0, 10)

# Tasks 

## 2. Find if there is anything about the df_ords AND df_prods that looks off or should be investigated further.

In [26]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### count for days_since_prior_order is less than other columns indicating missing values
#### order_dow is ranged from 0-6 indicating 7 days where index starts by 0
#### order_hour_of_day is within the range of 00:00 and 23:00 which is correct for 24 hour
#### there are 0 prior days on days_since_prior_order indicating some orders have been placed at the same day

In [63]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### abnormality on products table, is seen on the price column, where min value is 1.0 whereas max value is extremely high showing 99999 which is a sign on data inconsistency and should be further investigated. 

### 3. Check for mixed-type data in your df_ords dataframe.

In [28]:
for col in df_ords.columns.tolist():
  mixed_ords = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[mixed_ords]) > 0:
    print ("here are mixed type columns",(col))
else: 
    print ("no mixed type column found")

no mixed type column found


## 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

##### no mixed data was found on step 3

### 5. Run a check for missing values in your df_ords dataframe.
#### In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [29]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

##### there are 206209 missing records on the days_since_prior_order column

### 6. Address the missing values using an appropriate method
#### In a markdown cell, explain why you used your method of choice.

In [30]:
df_ords_prior_days_nan = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [31]:
df_ords_prior_days_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

~filling in empty cells in a column using fillna~

~df_ords['days_since_prior_order'].fillna(7, inplace=True)~

## REVISION: 
considering the high number of missing values (206209) in days_since_prior_order, it would be best to do nothing rather than replacing them because that would alter the data significantly, and there is a good reason for that much data being missing–first orders have no prior orders.

In [32]:
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

### 7. Run a check for duplicate values in your df_ords data

In [33]:
df_dup_ords = df_ords[df_ords.duplicated()]

In [33]:
df_dup_ords.shape

(0, 7)

#### no duplicated value is found - skipping task 8.

### 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [39]:
df_ords.to_csv(os.path.join(data_path, '02 Data','Prepared Data', 'orders_consistency_checked.csv'),index = False) 

In [40]:
df_ords.columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'order_dow',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

#### cleaned up data for products is already exported during the exercise